[16.0] fastapi: eases tests development

lmignon opened the pull request:
@marielejeune @sbidoul @AnizR This PR provides a base class to ease the development of your tests. Your feedback will be welcome

lmignon committed:
[IMP] fastapi: Add FastAPITransactionCase

This base class provides a method to ease writing tests for FastAPI routers and app

lmignon committed:
[IMP] fastapi: Refactor demo app to follow dev guidelines

AnizR commented on fastapi/README.rst:
```diff
@@ -1006,49 +1006,63 @@ normally provided by the normal processing of the request by the fastapi app.
+  from odoo.addons.fastapi.routers import demo_router
 
-  from odoo.tests.common import TransactionCase
+  class FastAPIDemoCase(FastAPITransactionCase):
```
I think that we should run the test with @tagged("post_install", "-at_install") or did it change?

marielejeune commented on fastapi/README.rst:
```diff
@@ -1006,49 +1006,63 @@ normally provided by the normal processing of the request by the fastapi app.
 .. code-block:: python
 
-  from functools import partial
+  from odoo.fastapi.tests.common import FastAPITransactionCase
```
```suggestion
  from odoo.addons.fastapi.tests.common import FastAPITransactionCase
```

marielejeune commented on fastapi/readme/USAGE.rst:
```diff
@@ -931,49 +931,63 @@ normally provided by the normal processing of the request by the fastapi app.
-With all these features, writing a test for the 'Hello world' route handler
-defined into the demo app is as simple as
+The fastapi addon provides a base class for the test cases that you can use to
+write your tests. This base class is **'odoo.fastapi.tests.common.FastAPITransactionCase'**.
```
```suggestion
write your tests. This base class is **'odoo.addons.fastapi.tests.common.FastAPITransactionCase'**.
```

marielejeune commented on fastapi/readme/USAGE.rst:
```diff
@@ -931,49 +931,63 @@ normally provided by the normal processing of the request by the fastapi app.
 .. code-block:: python
 
-  from functools import partial
+  from odoo.fastapi.tests.common import FastAPITransactionCase
```
```suggestion
  from odoo.addons.fastapi.tests.common import FastAPITransactionCase
```

marielejeune commented on fastapi/readme/USAGE.rst:
```diff
@@ -931,49 +931,63 @@ normally provided by the normal processing of the request by the fastapi app.
-      def tearDownClass(cls) -> None:
-          odoo_env_ctx.reset(cls._ctx_token)
-          cls.fastapi_demo_app._reset_app()
+  from odoo.fastapi.tests.common import FastAPITransactionCase
```
```suggestion
  from odoo.addons.fastapi.tests.common import FastAPITransactionCase
```

lmignon committed:
[IMP] fastapi: Prefer typing.Annotated to describe method paramters

lmignon committed:
[FIX] extendable:  Fix TestMixin class

Rename methods and avoid name collision

lmignon committed:
[IMP] extendable_fastapi: Add class to ease tests with fastapi and extendable

lmignon committed:
[IMP] fastapi: Always run tests at post-install

When loading the registry, Odoo takes rlock on the current thread to ensure the thead safety of the loading process. Nevertheless, dealing at the same time with lock on thread and the asyncio eventloop lead to the blockage of the eventloop. By running tests at post install, we avoid this kind of problem

AnizR reviewed (approved):
Code LGTM, I will try it on my module

sbidoul commented on extendable_fastapi/tests/common.py:
```diff
@@ -0,0 +1,21 @@
+
+    @classmethod
+    def tearDownClass(cls) -> None:
+        super().tearDownClass()
+        cls.reset_extendable_registry()
```
```suggestion
        cls.reset_extendable_registry()
        super().tearDownClass()
```

lmignon replied:
done

sbidoul commented on base_rest_demo/tests/common.py:
```diff
@@ -29,14 +29,18 @@ def setUpClass(cls):
+
+    @classmethod
+    def tearDownClass(cls):
+        super().tearDownClass()
+        cls.reset_extendable_registry()
```
```suggestion
        cls.reset_extendable_registry()
        super().tearDownClass()
```

to tear down in reverse order as setup?

lmignon replied:
done

lmignon commented:
Fixed and rebased after merge of #353

FrancoMaxime commented on fastapi/readme/USAGE.rst:
```diff
@@ -527,7 +538,7 @@ current request.
     )
     async def endpoint_app_info(
-        endpoint: FastapiEndpoint = Depends(fastapi_endpoint),  # noqa: B008
+        endpoint: Annotated[FastapiEndpoint, Depends(fastapi_endpoint)],  # noqa: B008
```
Is the # noqa: B008 still required? 
It is not used in other function declarations below

FrancoMaxime reviewed (commented):
LGTM

lmignon commented:
/ocabot merge patch
